sbi/utils/sbiutils.py

```diff
@@ -948,3 +951,171 @@ def mog_log_prob(
     exponent = -0.5 * utils.batched_mixture_vmv(precisions_pp, theta_minus_mean)
 
     return torch.logsumexp(weights + constant + log_det + exponent, dim=-1)
+
+
+transform_types = Optional[
+    Union[
+        torch.distributions.transforms.Transform,
+        torch.distributions.transforms.ComposeTransform,
+    ]
+]
+
+
+def get_log_abs_det_jacobian(
+    transform,
+    parameters_constrained: torch.Tensor,
+    parameters_unconstrained: torch.Tensor,
+) -> torch.Tensor:
+    """Return log_abs_det_jacobian over batch of parameters.
+
+    Take sum over second dimension if needed, e.g., when torch < 1.8 is used.
+
+    Args:
+        transform: transform applied to the parameters
+        parameters_constrained:
+        parameters_unconstrained:
+
+    Returns:
+        torch.Tensor: log_abs_det_jacobian value for each batch entry.
+    """
+
+    batch_size, dim_parameters = parameters_constrained.shape
+
+    vals = transform.log_abs_det_jacobian(
+        parameters_constrained, parameters_unconstrained
+    )
+
+    # For torch < 1.8 log_abs_det_jacobian is returned for each dimension.
+    if vals.ndim > 1 and vals.shape[1] == dim_parameters:
+        vals = vals.sum(-1)
+
+    assert (
+        vals.numel == batch_size
+    ), "Mismatch in batch size, took sum over whole batch?"
+
+    return vals
+
+
+def get_kde(
+    X: torch.Tensor,
+    bandwidth: Union[float, str] = "cv",
+    transform: transform_types = None,
+    verbose: bool = True,
+    sample_weight: Optional[np.ndarray] = None,
+) -> KernelDensity:
+    """Get KDE estimator with selected bandwidth
+
+    Args:
+        X: Samples
+        bandwidth: Bandwidth method
+        transform: Optional transform
```

Comment:
what is the transform doing here? Is this a constrained / unconstrained space thing? Can this be accessed via the API?